ark/type/src/parser/shared.ts

```diff
@@ -0,0 +1,114 @@
+import type { validateDefinition } from "./definition.js"
+import { Scanner } from "./string/shift/scanner.js"
+
+type parsedKey<result extends KeyParseResult> = result
+
+type OptionalTuple<value> = readonly [value, "?"]
+
+type KeyParseResult = {
+	kind: ParsedKeyKind
+	innerKey: string | symbol
+}
+type ParsedKeyKind = "required" | "optional" | "indexed"
+
+export type parsedEntry<result extends EntryParseResult> = result
+
+export type OptionalValue<value> =
+	| OptionalStringDefinition<value & string>
+	| OptionalTuple<value>
+
+export type EntryParseResult = { innerValue: unknown } & KeyParseResult
+
+export type OptionalStringDefinition<name extends string = string> = `${name}?`
+
+export type IndexedKey<def extends string = string> = `[${def}]`
+
+export type validateObjectValue<def, $, args> = def extends OptionalTuple<
+	infer value
+>
+	? readonly [validateObjectValueString<value, $, args>, "?"]
+	: validateObjectValueString<def, $, args>
+
+type validateObjectValueString<def, $, args> =
+	def extends OptionalStringDefinition<infer innerValue>
+		? `${validateDefinition<innerValue, $, args>}?`
+		: validateDefinition<def, $, args>
+
+const getInnerDataAndOptional = (
+	data: PropertyKey
+): { inner: PropertyKey; optional: boolean } =>
+	typeof data === "string" && data[data.length - 2] === Scanner.escapeToken
+		? { inner: `${data.slice(0, -2)}?`, optional: false }
+		: typeof data === "string"
+		? { inner: data.slice(0, -1), optional: true }
+		: { inner: data, optional: false }
+
+export const parseKeyValueEntry = (
```

Comment:
This should just return an `EntryParseResult`. In general we want to clean up these inlined-object types. If they're being returned from a function, they should be referencing a standalone type so they can be used, e.g. by the function calling them.